tests/worker_deployment_version_test.go

```diff
@@ -370,30 +353,119 @@ func (s *DeploymentVersionSuite) TestDrainageStatus_SetCurrentVersion_NoOpenWFs(
 	s.checkVersionDrainage(ctx, tv1, &deploymentpb.VersionDrainageInfo{
 		Status:          enumspb.VERSION_DRAINAGE_STATUS_DRAINED,
 		LastChangedTime: nil, // don't test this now
 		LastCheckedTime: nil, // don't test this now
-	})
-
-	// todo carly: not really sure how to check lastChangedTime / lastCheckedTime
+	}, true, false)
 }
 
 func (s *DeploymentVersionSuite) TestDrainageStatus_SetCurrentVersion_YesOpenWFs() {
-	// todo carly: test with open workflows on the draining version that then complete
-}
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+	tv1 := testvars.New(s).WithBuildIDNumber(1)
+	tv2 := testvars.New(s).WithBuildIDNumber(2)
+
+	// Start deployment workflow 1 and wait for the deployment version to exist
+	s.startVersionWorkflow(ctx, tv1)
+
+	// Start deployment workflow 2 and wait for the deployment version to exist
+	s.startVersionWorkflow(ctx, tv2)
+
+	// non-current deployments have never been used and have no drainage info
+	s.checkVersionDrainage(ctx, tv1, &deploymentpb.VersionDrainageInfo{}, false, false)
+	s.checkVersionDrainage(ctx, tv2, &deploymentpb.VersionDrainageInfo{}, false, false)
+
+	// SetCurrent tv1
+	_, err := s.FrontendClient().SetWorkerDeploymentCurrentVersion(ctx, &workflowservice.SetWorkerDeploymentCurrentVersionRequest{
+		Namespace:               s.Namespace().String(),
+		DeploymentName:          tv1.DeploymentSeries(),
+		Version:                 tv1.DeploymentVersionString(),
+		Identity:                tv1.ClientIdentity(),
+		IgnoreMissingTaskQueues: true,
+	})
+	s.Nil(err)
+
+	// both still nil since neither are draining
+	s.checkVersionDrainage(ctx, tv1, &deploymentpb.VersionDrainageInfo{}, false, false)
+	s.checkVersionDrainage(ctx, tv2, &deploymentpb.VersionDrainageInfo{}, false, false)
+
+	// start a pinned workflow on v1
+	run := s.startPinnedWorkflow(ctx, tv1)
+
+	// SetCurrent tv2 --> tv1 starts the child drainage workflow
+	_, err = s.FrontendClient().SetWorkerDeploymentCurrentVersion(ctx, &workflowservice.SetWorkerDeploymentCurrentVersionRequest{
+		Namespace:               s.Namespace().String(),
+		DeploymentName:          tv2.DeploymentSeries(),
+		Version:                 tv2.DeploymentVersionString(),
+		Identity:                tv2.ClientIdentity(),
+		IgnoreMissingTaskQueues: true,
+	})
+	s.Nil(err)
 
-func (s *DeploymentVersionSuite) TestDrainageStatus_SetRampingVersion_NoOpenWFs() {
-	// todo carly: test with open workflows on the draining version that then complete
+	// tv1 should now be "draining" for visibilityGracePeriod duration
+	s.checkVersionDrainage(ctx, tv1, &deploymentpb.VersionDrainageInfo{
+		Status:          enumspb.VERSION_DRAINAGE_STATUS_DRAINING,
+		LastChangedTime: nil, // don't test this now
+		LastCheckedTime: nil, // don't test this now
+	}, false, false)
+
+	// tv1 should still be "draining" for visibilityGracePeriod duration
+	s.checkVersionDrainage(ctx, tv1, &deploymentpb.VersionDrainageInfo{
```

Comment:
this doesn't wait for refresh period because Eventually will see the desired result in the first try.

To check that drainage happened after grace+drainage period, you can instead use the drainageChangedTime after you get Drained and make sure it was at least grace+drainage time after the time you v1 became non-current.

Reply:
The CI is almost passed, so I'm going to let this merge first, and then make this test change in a future PR